### ISSUE TITLE
Override method signature using typing.Self fails with reportIncompatibleMethodOverride

### ISSUE THREAD

**Issue by AllanDaemon** [microsoft/pyright#10488]
I cannot explain as well as the code itself.

This works on Mypy but in Pyight give the error `reportIncompatibleMethodOverride`. I believe this should be a correct code.

This is a simplified version from the original that preservers the error and context:
```python
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Self


from pathlib import Path
class RepoLocation(Path):
    """Mock of RepoLocation for the bug report"""
    @property
    def meta(self) -> Self:
        return self / "meta_file"


@dataclass
class SnapshotIndex:
    """Cheap class"""

    loc: RepoLocation
    time: datetime
    index: int

    # form a linked list of previous snapshots. SHOULDL BE from the same type,
    # i.e., type(a) == type(a.previous) for any a
    previous: Self | None

    @classmethod
    def from_loc(cls, snap_loc: RepoLocation, previous: Self | None = None) -> Self:
        time = datetime.fromisoformat(snap_loc.name)
        return cls(loc=snap_loc, time=time, index=-1, previous=previous)

    @classmethod
    def load_meta(cls, snap_loc: RepoLocation) -> dict:
        return {}  # In reality loads a lot of data here


@dataclass
class Snapshot(SnapshotIndex):
    """Expensive covariant"""

    meta: dict

    @classmethod
    def from_loc(cls, snap_loc: RepoLocation, previous: Self | None = None) -> Self:
        #   ^^^^^^^^ ERROR HERE due to this param ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        assert snap_loc.meta.is_file()
        time = datetime.fromisoformat(snap_loc.name)
        meta = cls.load_meta(snap_loc)
        return cls(loc=snap_loc, time=time, index=-1, previous=previous, meta=meta)
```

For instance, the `previous` is to create a linked list of the same type, so when calling `SnapshotIndex.from_loc()` previous should be a `SnapshotIndex`, and when calling `Snapshot.from_loc()` previous should be a `Snapshot`.

Pyright versions: Pylance 2025.5.1 (VS Code); pyright 1.1.401 (CLI)
Python version: 3.12
Mypy version: mypy 1.15.0 (compiled: yes)

```sh
$ mypy src/mcbkp/models/bug.py
Success: no issues found in 1 source file

$ pyright src/mcbkp/models/bug.py
/src/mcbkp/models/bug.py
  /src/mcbkp/src/mcbkp/models/bug.py:42:6 - error: Method "from_loc" overrides class "SnapshotIndex" in an incompatible manner
    Parameter 3 type mismatch: base parameter is type "SnapshotIndex | None", override parameter is type "Snapshot | None"
      Type "SnapshotIndex | None" is not assignable to type "Snapshot | None"
        Type "SnapshotIndex" is not assignable to type "Snapshot | None"
          "SnapshotIndex" is not assignable to "Snapshot"
          "SnapshotIndex" is not assignable to "None" (reportIncompatibleMethodOverride)
1 error, 0 warnings, 0 informations 

```

Maybe I'm incorrect about the correctness of the typing here. I 90% sure it's not, and the fact it works on Mypy gives more confidence about it, hence the bug report.

**Comment by erictraut** [comment]
Pyright is correct to generate an error here. This code violates the LSP (Liskov substitution principle).

Here's a demonstration of how this can lead to a crash at runtime. Note that none of the code below contains any type violations. The problem is due to the LSP violation.

```python
def test(s1: SnapshotIndex, s2: SnapshotIndex):
    s1.from_loc(RepoLocation(), s2)

r = RepoLocation()
s1 = Snapshot(r, datetime.now(), 0, None, {})
s2 = SnapshotIndex(r, datetime.now(), 0, None)

test(s1, s1)  # OK
test(s2, s2)  # Crash
```
